airbyte-integrations/connectors/destination-bigquery/src/main/kotlin/io/airbyte/integrations/destination/bigquery/migrators/BigqueryAirbyteMetaAndGenerationIdMigration.kt

```diff
@@ -0,0 +1,97 @@
+import io.airbyte.integrations.base.destination.typing_deduping.StreamConfig
+import io.airbyte.integrations.base.destination.typing_deduping.migrators.Migration
+import io.github.oshai.kotlinlogging.KotlinLogging
+
+class BigqueryAirbyteMetaAndGenerationIdMigration(private val bigquery: BigQuery) :
+    Migration<BigQueryDestinationState> {
+    private val logger = KotlinLogging.logger {}
+
+    override fun migrateIfNecessary(
+        destinationHandler: DestinationHandler<BigQueryDestinationState>,
+        stream: StreamConfig,
+        state: DestinationInitialStatus<BigQueryDestinationState>
+    ): Migration.MigrationResult<BigQueryDestinationState> {
+        if (!state.initialRawTableStatus.rawTableExists) {
+            // The raw table doesn't exist. No migration necessary. Update the state.
+            logger.info {
+                "Skipping airbyte_meta/generation_id migration for ${stream.id.originalNamespace}.${stream.id.originalName} because the raw table doesn't exist"
+            }
+            return Migration.MigrationResult(state.destinationState, false)
+        }
+
+        val rawTable = bigquery.getTable(TableId.of(stream.id.rawNamespace, stream.id.rawName))
+        // if the schema is null, then we have bigger problems
+        val rawFields = rawTable.getDefinition<StandardTableDefinition>().schema!!.fields
+        val hasMeta = rawFields.any { it.name == JavaBaseConstants.COLUMN_NAME_AB_META }
+        if (hasMeta) {
+            // We've already executed the migration. Do nothing here.
+            logger.info {
+                "Skipping airbyte_meta/generation_id migration for ${stream.id.originalNamespace}.${stream.id.originalName} because the table already has the columns"
+            }
+            return Migration.MigrationResult(state.destinationState, false)
+        }
+
+        logger.info {
+            "Executing airbyte_meta/generation_id migration for ${stream.id.originalNamespace}.${stream.id.originalName}"
+        }
+
+        // Define the new columns we're adding.
+        // Add meta to the raw table
+        val rawMetaField =
+            Field.of(JavaBaseConstants.COLUMN_NAME_AB_META, StandardSQLTypeName.STRING)
+        // And add generation ID to raw+final tables.
+        val generationIdField =
+            Field.of(JavaBaseConstants.COLUMN_NAME_AB_GENERATION_ID, StandardSQLTypeName.INT64)
+
+        // The way bigquery does an "alter table" is by just setting the entire table definition
+        // (unless you want to write actual SQL, of course).
+        // This adds the column as NULLABLE.
+        val newRawTable =
+            rawTable
+                .toBuilder()
+                .setDefinition(
+                    StandardTableDefinition.of(
+                        Schema.of(rawFields + rawMetaField + generationIdField)
+                    )
+                )
+                .build()
+        newRawTable.update()
+
+        if (state.isFinalTablePresent) {
```

Comment:
afaik yeah. definitely a little weird. There's probably technically edge cases that it doesn't handle (e.g. if the raw table alter succeeds, then the final table alter fails - should we detect that in the next run?)

Reply:
in redshift, we let it happen using softReset iirc. we are avoiding that here ?

Reply:
did we need to change the final table at all in redshift? I thought that was purely a raw table change to add airbyte_meta, since the final table already had airbyte_meta?

Reply:
yeah, we're explicitly setting softReset=false here https://github.com/airbytehq/airbyte/blob/d31f0dd7a7afbc503ef67c224d3f55b1860e3cde/airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/typing_deduping/RedshiftRawTableAirbyteMetaMigration.kt#L74-L77

Reply:
oh you are right, final table had meta as part of v1v2 migration doing soft reset.